ops/storage.py

```diff
@@ -0,0 +1,293 @@
+        self._db.execute("DELETE FROM snapshot WHERE handle=?", (handle_path,))
+
+    def save_notice(self, event_path, observer_path, method_name):
+        self._db.execute('INSERT INTO notice VALUES (NULL, ?, ?, ?)',
+                         (event_path, observer_path, method_name))
+
+    def drop_notice(self, event_path, observer_path, method_name):
+        self._db.execute('''
+            DELETE FROM notice
+             WHERE event_path=?
+               AND observer_path=?
+               AND method_name=?
+            ''', (event_path, observer_path, method_name))
+
+    def notices(self, event_path):
+        if event_path:
+            c = self._db.execute('''
+                SELECT event_path, observer_path, method_name
+                  FROM notice
+                 WHERE event_path=?
+                 ORDER BY sequence
+                ''', (event_path,))
+        else:
+            c = self._db.execute('''
+                SELECT event_path, observer_path, method_name
+                  FROM notice
+                 ORDER BY sequence
+                ''')
+        while True:
+            rows = c.fetchmany()
+            if not rows:
+                break
+            for row in rows:
+                yield tuple(row)
+
+
+class JujuStorage:
+    """"Storing the content tracked by the Framework in Juju.
+
+    This uses :class:`_JujuStorageBackend` to interact with state-get/state-set
+    as the way to store state for the framework and for components.
+    """
+
+    NOTICE_KEY = "#notices#"
+
+    def __init__(self, backend: '_JujuStorageBackend' = None):
+        self._backend = backend
+        if backend is None:
+            self._backend = _JujuStorageBackend()
+
+    def close(self):
+        return
+
+    def commit(self):
+        return
+
+    def save_snapshot(self, handle_path: str, snapshot_data: typing.Any) -> None:
+        self._backend.set(handle_path, snapshot_data)
+
+    def load_snapshot(self, handle_path):
+        try:
+            content = self._backend.get(handle_path)
+        except KeyError:
+            raise NoSnapshotError(handle_path)
+        return content
+
+    def drop_snapshot(self, handle_path):
+        self._backend.delete(handle_path)
+
+    def save_notice(self, event_path: str, observer_path: str, method_name: str):
+        notice_list = self._load_notice_list()
+        notice_list.append([event_path, observer_path, method_name])
+        self._save_notice_list(notice_list)
+
+    def drop_notice(self, event_path: str, observer_path: str, method_name: str):
+        notice_list = self._load_notice_list()
+        notice_list.remove([event_path, observer_path, method_name])
+        self._save_notice_list(notice_list)
+
+    def notices(self, event_path: str):
+        notice_list = self._load_notice_list()
+        if event_path:
+            notice_list = list(filter(lambda r: r[0] == event_path, notice_list))
+        for row in notice_list:
```

Comment:
If we're just iterating it here, we don't really need to build a list above

Reply:
ditto :-D

Reply:
So I think a bigger question is whether we want to store notices by a prefix so that we don't have to load them all.
Though IMO, we should push that you don't have 100s of events that haven't been handled.
And we have 2 code paths
1) Iterate all events for _reemit() all deferred
2) Iterate only a specific event's notices.

I think using a structure of all notices in one list and just filtering them is fine.

I do wonder why have the notice list as a comprehension vs just 'for row in notice_list, if r[0] !=, continue'. I'll tweak.